source/server/admin/logs_handler.cc

```diff
@@ -51,64 +63,123 @@ Http::Code LogsHandler::handlerReopenLogs(absl::string_view, Http::ResponseHeade
+    if (!level_to_use.has_value()) {
+      return std::make_pair(false, "unknown log level");
     }
-  }
 
-  if (level_to_use == std::numeric_limits<size_t>::max()) {
-    return false;
+    changeAllLogLevels(level_to_use);
+    return std::make_pair(true, "");
   }
 
-  if (!Logger::Context::useFancyLogger()) {
-    // Now either change all levels or a single level.
-    if (name == "level") {
-      ENVOY_LOG(debug, "change all log levels: level='{}'", level);
-      for (Logger::Logger& logger : Logger::Registry::loggers()) {
-        logger.setLevel(static_cast<spdlog::level::level_enum>(level_to_use));
+  // Build a map of name:level pairs, a few allocations is ok here since it's
+  // not common to call this function at a high rate.
+  LogLevelMap name_levels;
+
+  if (key == "paths") {
+    // Bulk change log level by name:level pairs, separated by comma.
+    std::vector<absl::string_view> pairs = absl::StrSplit(value, ',', absl::SkipWhitespace());
+    for (const auto& name_level : pairs) {
+      std::pair<absl::string_view, absl::string_view> name_level_pair =
+          absl::StrSplit(name_level, absl::MaxSplits(':', 1), absl::SkipWhitespace());
+      auto [name, level] = name_level_pair;
+      if (name.empty() || level.empty()) {
+        return std::make_pair(false, "empty log name or empty log level");
       }
-    } else {
-      ENVOY_LOG(debug, "change log level: name='{}' level='{}'", name, level);
-      Logger::Logger* logger_to_change = nullptr;
-      for (Logger::Logger& logger : Logger::Registry::loggers()) {
-        if (logger.name() == name) {
-          logger_to_change = &logger;
-          break;
-        }
+
+      auto level_to_use = parseLevel(level);
+      if (!level_to_use.has_value()) {
+        return std::make_pair(false, "unknown log level");
       }
 
-      if (!logger_to_change) {
-        return false;
+      name_levels[name] = *level_to_use;
+    }
+  } else {
+    // Change particular log level by name.
+    auto level_to_use = parseLevel(value);
+    if (!level_to_use.has_value()) {
+      return std::make_pair(false, "unknown log level");
+    }
+
+    name_levels[name] = *level_to_use;
+  }
+
+  return changeLogLevels(name_levels, errors);
+}
+
+void LogsHandler::changeAllLogLevels(spdlog::level::level_enum level) {
+  if (!Logger::Context::useFancyLogger()) {
+    ENVOY_LOG(debug, "change all log levels: level='{}'", spdlog::level::level_string_views[level]);
+    Logger::Registry::setLogLevel(*level_to_use);
+  } else {
+    // Level setting with Fancy Logger.
+    FANCY_LOG(info, "change all log levels: level='{}'", spdlog::level::level_string_views[level]);
+    getFancyContext().setAllFancyLoggers(*level_to_use);
+  }
+}
+
+std::pair<bool, std::string> LogsHandler::changeLogLevels(const LogLevelMap& changes) {
+  if (!Logger::Context::useFancyLogger()) {
+    // Build a map of name to logger.
+    absl::flat_hash_map<std::string, Logger::Logger*> loggers;
```

Comment:
you don't need to make this new map either. You can make this efficient using the one map you already have (`changes`).